store/PoolInstancesSlice/index.tsx

```diff
@@ -148,67 +125,122 @@ export const createPoolsInstancesSlice: StateSlice<IPoolsInstancesSlice> = (set,
-                const settlementTokenBalance = new BigNumber(ethers.utils.formatUnits(balances[2], decimals));
-
-                console.debug('Balances', {
-                    shortTokenBalance,
-                    longTokenBalance,
-                    settlementTokenBalance,
-                });
-
-                get().setTokenBalances(pool.address, {
-                    shortTokenBalance,
-                    longTokenBalance,
-                    settlementTokenBalance,
-                });
-            })
-            .catch((err) => {
-                console.error('Failed to fetch token balances', err);
-            });
-        fetchAggregateBalance(provider, account, pool.committer.address, decimals)
+        const uniqueSettlementTokensMap: Record<string, string[]> = Object.values(get().pools).reduce(
+            (
+                o,
+                {
+                    poolInstance: {
+                        address,
+                        settlementToken: { address: tokenAddress },
+                    },
+                },
+            ) => {
+                o[tokenAddress] = o[tokenAddress] || [];
+                o[tokenAddress].push(address);
+                return o;
+            },
+            {} as Record<string, string[]>,
+        );
+        const relevantSettlementTokens = Object.keys(
+            pools_.reduce((o, pool) => {
+                const settlementToken = get().pools[pool]?.poolInstance?.settlementToken?.address;
+                if (settlementToken) {
+                    o[settlementToken] = true;
+                }
+                return o;
+            }, {} as Record<string, boolean>),
+        );
+        fetchTokenBalances(relevantSettlementTokens, provider, account)
             .then((balances) => {
-                console.debug('Pending balances', {
-                    longTokens: balances.longTokens.toNumber(),
-                    shortTokens: balances.shortTokens.toNumber(),
-                    settlementTokens: balances.settlementTokens.toNumber(),
+                console.count('Fetched settlement token balances');
+                set((state) => {
+                    balances.map((settlementTokenBalance_, index) => {
+                        const pools = uniqueSettlementTokensMap[relevantSettlementTokens[index]];
+                        // pool[0] must exist otherwise the entry would not exist in uniqueSettlementTokens
+                        const decimals = state.pools[pools[0]].poolInstance.settlementToken.decimals;
+                        const settlementTokenBalance = new BigNumber(
+                            ethers.utils.formatUnits(settlementTokenBalance_, decimals),
+                        );
+                        uniqueSettlementTokensMap[relevantSettlementTokens[index]].map((pool) => {
+                            state.pools[pool].userBalances.settlementToken.balance = settlementTokenBalance;
+                        });
```

Comment:
not a huge fan of using `.map` to perform looping without actually using the return values but it works

Reply:
updated to forEach